kernel/mmanutils.c

```diff
@@ -517,168 +528,141 @@ int myst_get_free_ram(size_t* size)
-    size_t size;
-    off_t offset; // file offset
-    char* pathname;
-    int prot;
-};
-
-static myst_process_mapping_t* _mappings;
-static myst_spinlock_t _mappings_lock;
-
-/* uncomment the following line to perform memchecks on process mappings */
-// #define MEMCHECK_PROCESS_MAPPINGS
-
-static void* _calloc(size_t nmemb, size_t size)
-{
-#ifdef MEMCHECK_PROCESS_MAPPINGS
-    return calloc(nmemb, size);
-#else
-    return myst_calloc(nmemb, size);
-#endif
-}
-
-static void _free(void* ptr)
-{
-#ifdef MEMCHECK_PROCESS_MAPPINGS
-    return free(ptr);
-#else
-    return myst_free(ptr);
-#endif
-}
-
-static void* _strdup(const char* s)
-{
-#ifdef MEMCHECK_PROCESS_MAPPINGS
-    return strdup(s);
-#else
-    return myst_strdup(s);
-#endif
-}
-
-/* keep track of mappings made by this process */
-int myst_register_process_mapping(
-    pid_t pid,
-    void* addr,
-    size_t size,
-    int fd,
-    off_t offset,
-    int prot)
+/* release mappings owned by the given process */
+int myst_release_process_mappings(pid_t pid)
 {
     int ret = 0;
-    myst_process_mapping_t* m = NULL;
-    struct locals
-    {
-        char realpath[PATH_MAX];
-    }* locals = NULL;
 
-    if (pid < 0 || !addr || (addr == (void*)-1) || !size)
+    if (pid <= 0)
         ERAISE(-EINVAL);
 
-    if (!(m = _calloc(1, sizeof(myst_process_mapping_t))))
-        ERAISE(-ENOMEM);
+    {
+        uint8_t* addr = (uint8_t*)_mman.map;
+        size_t length = ((uint8_t*)_mman.end) - addr;
+        size_t index;
+        vectors_t v = _get_vectors();
 
-    m->pid = pid;
-    m->addr = addr;
-    m->size = size;
+        assert(v.pids_count == v.fdmappings_count);
 
-    /* If file mapping, save pathname, file offset and prot */
-    if (fd >= 0)
-    {
-        int ret = 0;
-        myst_file_t* file = NULL;
-        myst_fs_t* fs = NULL;
-        myst_fdtable_t* fdtable = myst_fdtable_current();
-
-        if (!(locals = malloc(sizeof(struct locals))))
-            ERAISE(-ENOMEM);
-
-        /* Get file path */
-        ECHECK(myst_fdtable_get_file(fdtable, fd, &fs, &file));
-        ECHECK((*fs->fs_realpath)(
-            fs, file, locals->realpath, sizeof(locals->realpath)));
-
-        /* Copy file path, offset and prot to mappings struct */
-        if (!(m->pathname = _strdup(locals->realpath)))
-            ERAISE(-ENOMEM);
-        m->offset = offset;
-        m->prot = prot;
-    }
+        ECHECK(myst_round_up(length, PAGE_SIZE, &length));
+        ECHECK((index = _get_page_index(addr, length)));
+        size_t count = length / PAGE_SIZE;
 
-    myst_spin_lock(&_mappings_lock);
-    {
-        m->next = _mappings;
-        _mappings = m;
-        m = NULL;
-    }
-    myst_spin_unlock(&_mappings_lock);
+        assert(index < v.pids_count);
+        assert(index + count <= v.pids_count);
 
-done:
+        myst_rspin_lock(&_mman.lock);
+        {
+            for (size_t i = index; i < index + count;)
+            {
+                if (v.pids[i] == (uint32_t)pid)
```

Comment:
I think this is too vague for a check.
Lets say a while file is mapped to a file. Then if part of the mapping is unmapped (code in the else) then what happens if the same process gets this newly unmapped page? It will release it incorrectly.